### PR TITLE
build: migrate web build to Vite

### LINKED ISSUE #3808: [web-infra] Switch to Vite over esbuild

### PR DESCRIPTION
**Pull request by Parker-Stafford**
resolves #3808 

Migrates the web application to use Vite for development and production builds. This allows for a better developer experience due to HMR and fast build times and splitting up the javascript bundle to speed up page loads.

Testing:

- [x] sagemaker notebooks
- [x] colab notebooks
- [x] docker containers
- [x] local devlopment

### CODE REVIEW

File: src/phoenix/server/templates/index.html [hunk 1]
```diff
@@ -21,28 +21,38 @@
...
+      import RefreshRuntime from 'http://localhost:5173/@react-refresh'
+      RefreshRuntime.injectIntoGlobalHook(window)
+      window.$RefreshReg$ = () => {}
+      window.$RefreshSig$ = () => (type) => type
+      window.__vite_plugin_react_preamble_installed__ = true
+    </script>
+    <script type="module" src="http://localhost:5173/@vite/client"></script>
+    <script type="module" src="http://localhost:5173/index.tsx"></script>
+   
+    <!-- <script type="module" src="{{basename}}/index.js"></script> -->
```
Comment (Parker-Stafford) [inline src/phoenix/server/templates/index.html:56]: todo: add this back and conditionally run the above 3 scripts in development

File: app/src/index.tsx [hunk 2]
```diff
@@ -1,7 +1,10 @@
 import React from "react";
+// Rest of your app code
 import ReactDom from "react-dom/client";
 
-import { App } from "./App";
+import "vite/modulepreload-polyfill";
```
Comment (mikeldking) [inline app/src/index.tsx:5]: Can you link why this is needed? https://vitejs.dev/config/build-options#build-modulepreload

File: app/package.json [hunk 3]
```diff
@@ -40,6 +40,7 @@
     "recharts": "^2.10.4",
     "relay-runtime": "^16.2.0",
     "remark-gfm": "^4.0.0",
+    "rollup-plugin-visualizer": "^5.12.0",
```
Comment (mikeldking) [inline app/package.json:43]: Dev deps

File: app/src/pages/datasets/DatasetsPage.tsx [hunk 4]
```diff
@@ -141,7 +141,7 @@ function CreateDatasetActionMenu({
   return (
     <>
       <ActionMenu
-        buttonText="Create Dataset"
+        buttonText="Create Dataset test"
```
Comment (mikeldking) [inline app/src/pages/datasets/DatasetsPage.tsx:144]:
```suggestion
        buttonText="Create Dataset"
```

**Comment by Parker-Stafford** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Parker-Stafford** [comment]
recheck

File: app/package.json [hunk 5]
```diff
@@ -62,6 +62,7 @@
     "@types/three": "^0.149.0",
     "@typescript-eslint/eslint-plugin": "^7.16.1",
     "@typescript-eslint/parser": "^7.16.1",
+    "@vitejs/plugin-react": "^4.3.1",
     "cpy-cli": "^5.0.0",
     "esbuild": "^0.19.12",
```
Comment (mikeldking) [inline app/package.json:67]: file a follow-up to just use vite or?

File: app/package.json [hunk 6]
```diff
@@ -75,37 +76,36 @@
...
-    "clean": "rimraf ../src/phoenix/server/static",
-    "build": "pnpm run clean && pnpm run build:static & pnpm run build:relay && pnpm run build:ts",
-    "build:static": "cpy ./static ../src/phoenix/server",
-    "build:ts": "node ./esbuild.config.mjs",
+    "build": "pnpm run build:relay && vite build",
     "build:relay": "relay-compiler",
-    "watch": "./esbuild.config.mjs dev",
     "test": "jest --config ./jest.config.js",
-    "dev": "pnpm run dev:server:demo & pnpm run build:static && pnpm run watch",
+    "dev": " pnpm run dev:server:demo & pnpm run build:relay && vite",
```
Comment (mikeldking) [inline app/package.json:90]:
```suggestion
    "dev": "pnpm run dev:server:demo && vite",
```

File: app/src/index.tsx [hunk 7]
```diff
@@ -1,7 +1,14 @@
...
 
-import { App } from "./App";
+/**
+ * This is needed to support modulepreload in Vite when using a non HTML custom entrypoint (i.e., this file).
+ * We do this because our index.html is served by the Phoenix server
+ * @see https://vitejs.dev/config/build-options#build-modulepreload
+ */
+import "vite/modulepreload-polyfill";
+
+import App from "./App";
```
Comment (mikeldking) [inline app/src/index.tsx:11]: curious why this needed to be switched to default export?

File: src/phoenix/server/app.py [hunk 8]
```diff
@@ -103,6 +104,8 @@ class AppConfig(NamedTuple):
     min_dist: float
     n_neighbors: int
     n_samples: int
+    is_development: bool
+    manifest_json_path: Path
```
Comment (mikeldking) [inline src/phoenix/server/app.py:108]: manifest is kinda a generic name - I'd either document it or call it the app_manifest or web_manifest?

File: app/package.json (hunk 5 above)
Comment (Parker-Stafford) [reply app/package.json:67]: oh i can remove and test again

File: src/phoenix/server/app.py [hunk 9]
```diff
@@ -112,8 +115,16 @@ class Static(StaticFiles):
...
         self._app_config = app_config
+        self.manifest = self._load_manifest()
         super().__init__(**kwargs)
 
+    def _load_manifest(self) -> Dict[str, Any]:
+        try:
+            with open(self._app_config.manifest_json_path, "r") as f:
+                return cast(Dict[str, Any], json.load(f))
+        except (FileNotFoundError, json.JSONDecodeError) as e:
+            raise e
```
Comment (mikeldking) [inline src/phoenix/server/app.py:126]: this can be cached https://docs.python.org/3/library/functools.html - maybe chat with some of the python folks

Comment (mikeldking) [inline src/phoenix/server/app.py:126]: any reason not to just raise? What's the need for the try catch?

File: src/phoenix/server/app.py (hunk 8 above)
Comment (Parker-Stafford) [reply src/phoenix/server/app.py:108]: good call

File: src/phoenix/server/app.py (hunk 9 above)
Comment (Parker-Stafford) [reply src/phoenix/server/app.py:126]: this was just cause i don't know python will update

File: app/src/index.tsx (hunk 7 above)
Comment (Parker-Stafford) [reply app/src/index.tsx:11]: ahh it probably doesn't in the early days i saw something that said that wa causing issues, I don't think it actually was the cause can switch back and see

File: app/package.json (hunk 6 above)
Comment (mikeldking) [reply app/package.json:89]: I don't think I fully follow but we can leave it for now. Can you remove the whitespace?

File: src/phoenix/server/templates/index.html [hunk 10]
```diff
@@ -1,3 +1,43 @@
...
+      {% endif %}
+    {% endif %}
+    
+    {% if chunk.file.endswith('.js') and chunk.file not in rendered_files %}
+      <link rel="modulepreload" href="{{ basename }}/{{ chunk.file }}">
+      {% set _ = rendered_files.append(chunk.file) %}
+    {% endif %}
+  {% endif %}
+{% endmacro %}
+
```
Comment (mikeldking) [inline src/phoenix/server/templates/index.html:40]: there's a bunch of whitespace here. It's not a huge deal but kinda confusing when inspecting the page.

File: src/phoenix/server/templates/index.html [hunk 11]
```diff
@@ -11,7 +51,6 @@
     <meta property="og:description" content="AI Observability & Evaluation" />
     <meta property="og:image" content="https://raw.githubusercontent.com/Arize-ai/phoenix-assets/main/images/socal/social-preview-horizontal.jpg" />
     <meta name="theme-color" content="#ffffff" />
-    <link rel="stylesheet" src="{{basename}}/index.css"></link>
```
Comment (mikeldking) [inline src/phoenix/server/templates/index.html:14]: why is the scripts not in the HEAD?

File: src/phoenix/server/templates/index.html [hunk 12]
```diff
@@ -21,28 +60,40 @@
...
+        RefreshRuntime.injectIntoGlobalHook(window)
+        window.$RefreshReg$ = () => {}
+        window.$RefreshSig$ = () => (type) => type
+        window.__vite_plugin_react_preamble_installed__ = true
+      </script>
+      <script type="module" src="http://localhost:5173/@vite/client"></script>
+      <script type="module" src="http://localhost:5173/index.tsx"></script>
+    {% else %}
+      {{ render_chunk('index.tsx') }}
+    {% endif %}
```
Comment (mikeldking) [inline src/phoenix/server/templates/index.html:97]: why not place these in the head?

File: app/.gitignore [hunk 13]
```diff
@@ -1,2 +1,3 @@
 node_modules
-dist
\ No newline at end of file
+dist
+stats.html
```
Comment (mikeldking) [inline app/.gitignore:3]: direct path? Seems a bit generic.

File: src/phoenix/server/app.py [hunk 14]
```diff
@@ -114,6 +118,22 @@ def __init__(self, *, app_config: AppConfig, **kwargs: Any):
...
+    @cached_property
+    def _web_manifest(self) -> Dict[str, Any]:
+        try:
+            with open(self._app_config.web_manifest_path, "r") as f:
+                return cast(Dict[str, Any], json.load(f))
+        except FileNotFoundError as e:
+            if self._app_config.is_development:
+                logger.warning(
+                    f"Web manifest not found at {self._app_config.web_manifest_path}"
+                )
```
Comment (mikeldking) [inline src/phoenix/server/app.py:130]: why do we need a waring? seems fine to skip

File: src/phoenix/server/templates/index.html (hunk 12 above)
Comment (mikeldking) [reply src/phoenix/server/templates/index.html:97]: nbd.

**Review by mikeldking — APPROVED** [review]
LGTM

File: src/phoenix/server/templates/index.html (hunk 11 above)
Comment (Parker-Stafford) [reply src/phoenix/server/templates/index.html:14]: moved css into head and kept js in body to not block page loading on js bundles

File: src/phoenix/server/templates/index.html (hunk 12 above)
Comment (Parker-Stafford) [reply src/phoenix/server/templates/index.html]: moved css into head and kept js in body to not block page loading on js bundles

File: src/phoenix/server/templates/index.html (hunk 10 above)
Comment (Parker-Stafford) [reply src/phoenix/server/templates/index.html]: fixed!

File: app/.gitignore (hunk 13 above)
Comment (Parker-Stafford) [reply app/.gitignore:3]: this is the gitignore nested under app so it has an implied nested path - it's what gets generated by vite dev i can look into nesting it via the visualizer plugin config

File: src/phoenix/server/app.py (hunk 14 above)
Comment (Parker-Stafford) [reply src/phoenix/server/app.py]: removed

File: app/package.json [hunk 15]
```diff
@@ -75,37 +75,37 @@
...
-    "clean": "rimraf ../src/phoenix/server/static",
-    "build": "pnpm run clean && pnpm run build:static & pnpm run build:relay && pnpm run build:ts",
+    "build": "pnpm run build:relay && vite build",
     "build:static": "cpy ./static ../src/phoenix/server",
-    "build:ts": "node ./esbuild.config.mjs",
     "build:relay": "relay-compiler",
-    "watch": "./esbuild.config.mjs dev",
     "test": "jest --config ./jest.config.js",
-    "dev": "pnpm run dev:server:demo & pnpm run build:static && pnpm run watch",
+    "dev": " pnpm run dev:server:demo & pnpm run build:static && pnpm run build:relay && vite",
```
Comment (mikeldking) [inline app/package.json:90]:
```suggestion
    "dev": "pnpm run dev:server:demo & pnpm run build:static && pnpm run build:relay && vite",
```

**Pull request merged**